File: src/main/java/org/example/expert/aop/ManagerReqLogAspect.java
package org.example.expert.aop;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.example.expert.aop.dto.ManagerLogContext;
import org.example.expert.exception.InvalidRequestException;
import org.example.expert.domain.log.manager.ManagerLogMessage;
import org.example.expert.domain.log.manager.ManagerReqStatus;
import org.example.expert.service.ManagerReqLogService;
import org.springframework.stereotype.Component;

@Aspect
@Component
@RequiredArgsConstructor
@Slf4j
public class ManagerReqLogAspect {
    private static final String MANAGER_SAVE_POINTCUT =
            "execution(* org.example.expert.domain.manager.service.ManagerService.saveManager(..))";

    private final ManagerReqLogService logService;

    @Around(MANAGER_SAVE_POINTCUT)
    public Object logManagerAssignment(ProceedingJoinPoint joinPoint) throws Throwable{
        ManagerLogContext managerLogContext = ManagerLogContext.validateAndCreate(joinPoint.getArgs());
        try{
            Object result = joinPoint.proceed();
            String message = ManagerLogMessage.ASSIGNED_SUCCESS.format(managerLogContext.getRequest().getTargetUserId(), managerLogContext.getTodoId());
            saveLog(managerLogContext, message, ManagerReqStatus.SUCCESS, null);
            return result;
        }catch (InvalidRequestException e) {
            String message = ManagerLogMessage.ASSIGNED_FAIL.format(e.getMessage(), managerLogContext.getTodoId(), managerLogContext.getRequest().getTargetUserId());
            saveLog(managerLogContext, message, ManagerReqStatus.FAIL, e);
            throw e;
        }catch(Exception e){
            String message = ManagerLogMessage.INTERNAL_FAIL.format(e.getMessage(), managerLogContext.getTodoId(), managerLogContext.getRequest().getTargetUserId());
            saveLog(managerLogContext, message, ManagerReqStatus.FAIL, e);
            throw e;
        }
    }

    private void saveLog (ManagerLogContext context, String message, ManagerReqStatus status, Exception originalException) {
        try {
            logService.saveLog(context.getRequest(), status, context.getUser(), context.getTodoId(), message);
        } catch (Exception logError) {
            if (originalException == null) {
                //비즈니스는 성공적으로 수행됐으나 로그 저장에 실패한 경우
                log.warn(ManagerLogMessage.LOG_SAVE_FAIL_WITH_ERROR.getFormat(), logError.getMessage(), logError);
            } else {
                log.warn(ManagerLogMessage.LOG_SAVE_FAIL_WITH_ERROR.getFormat(), originalException.getMessage(), logError.getMessage(), logError);
            }
        }
    }
}

File: src/main/java/org/example/expert/aop/ManagerReqLogAspect.kt
package org.example.expert.aop

import org.aspectj.lang.JoinPoint
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.example.expert.aop.dto.ManagerLogContext
import org.example.expert.domain.log.manager.ManagerLogMessage
import org.example.expert.domain.log.manager.ManagerReqStatus
import org.example.expert.exception.CustomApiException
import org.example.expert.exception.ErrorCode
import org.example.expert.service.ManagerReqLogService
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

@Aspect
@Component
class ManagerReqLogAspect(
    private val logService: ManagerReqLogService
) {
    companion object{
        private val log = LoggerFactory.getLogger(this::class.java)
        private const val MANAGER_LOG_POINTCUT: String =
            "execution(* org.example.expert.service.ManagerService.saveManager(..))"
    }
    @Around(MANAGER_LOG_POINTCUT)
    fun logManagerAssignment(joinPoint: ProceedingJoinPoint): Any {
        val managerLogContext = ManagerLogContext.validateAndCreate(joinPoint.args)
        try {
            val result = joinPoint.proceed()
            val message = ManagerLogMessage.ASSIGNED_SUCCESS.format(
                managerLogContext.todoId,
                managerLogContext.request.targetUserId
            )
            saveLog(managerLogContext, message, ManagerReqStatus.SUCCESS, null)
            return result
        } catch (e: CustomApiException){
            val message =
                ManagerLogMessage.ASSIGNED_FAIL.format(
                    e.errorCode.message,
                    managerLogContext.todoId,
                    managerLogContext.request.targetUserId)
            saveLog(managerLogContext, message, ManagerReqStatus.FAIL, e)
            throw e
        } catch (e: Exception){
            val message = ManagerLogMessage.INTERNAL_FAIL.format(
                e.message?: "알 수 없는 오류",
                managerLogContext.todoId,
                managerLogContext.request.targetUserId
            )
            saveLog(managerLogContext, message, ManagerReqStatus.FAIL, e)
            throw e
        }
    }

    private fun saveLog(context: ManagerLogContext,
                        message: String,
                        status: ManagerReqStatus,
                        originalException: Exception?){
        try{
            logService.saveLog(
                managerCreateReqDto = context.request,
                status = status,
                user = context.user,
                todoId = context.todoId,
                message = message
            )
        }catch (saveError: Exception){
            if(originalException == null){
                log.warn(
                    ManagerLogMessage.LOG_SAVE_FAIL_WITH_SUCCESS.format(
                        saveError.message?: "알 수 없는 오류"
                    ),
                    saveError)
            } else{
                log.warn(
                    ManagerLogMessage.LOG_SAVE_FAIL_WITH_ERROR.format(
                        originalException.message?: "알 수 없는 오류",
                        saveError.message?: "알 수 없는 오류"
                    ),
                    saveError)
            }
        }
    }
}